src/query/query_parser/logical_ast.rs

```diff
@@ -39,6 +39,33 @@ impl LogicalAst {
             LogicalAst::Boost(Box::new(self), boost)
         }
     }
+
+    pub fn simplify(self) -> LogicalAst {
+        match self {
+            LogicalAst::Clause(clauses) => {
+                let mut new_clauses: Vec<(Occur, LogicalAst)> = Vec::new();
+
+                for (occur, sub_ast) in clauses {
+                    let simplified_sub_ast = sub_ast.simplify();
+
+                    // If clauses below have the same `Occur`, we can pull them up
+                    match simplified_sub_ast {
+                        LogicalAst::Clause(sub_clauses)
+                            if sub_clauses.iter().all(|(o, _)| *o == occur) =>
```

Comment:
What if Occur is Not?

Reply:
yes good catch, I excluded it and added a test. There are more optimization we can do here to flatten clauses later on